toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1120,6 +1121,15 @@ func (g *PkgGraph) CreateSubGraph(rootNode *PkgNode) (subGraph *PkgGraph, err er
 	return
 }
 
+// IsSRPMAvailablePMC checks if an SRPM is prebuilt, returning true if so along with a slice of corresponding prebuilt RPMs.
+// The function will lock 'graphMutex' before performing the check if the mutex is not nil.
+func IsSRPMAvailablePMC(srpmPath string, pkgGraph *PkgGraph, graphMutex *sync.RWMutex, outDir, tmpDir, workerTar, existingRpmsDir, existingToolchainRpmsDir string, usePreviewRepo bool, repoFiles []string, disableUpstreamRepos bool, tlsClientKey string, tlsClientCert string) (isPrebuilt bool, expectedFiles, missingFiles []string) {
+	expectedRpmNodes := NodesProvidedBySRPM(srpmPath, pkgGraph, graphMutex)
+	isPrebuilt, missingFiles = findAllRPMSinPMC(expectedRpmNodes, outDir, tmpDir, workerTar, existingRpmsDir, existingToolchainRpmsDir, usePreviewRepo, repoFiles, disableUpstreamRepos, tlsClientKey, tlsClientCert)
+//	logger.Log.Tracef("Missing RPMs from %s: %v", srpmPath, missingFiles)
+	return
+}
+
```

Comment:
I think we're re-inventing the wheel here a bit and overcomplicating the code:

1. We already have a function to convert and SRPM to a list of built packages. It's the `pkggraph.rpmsProvidedBySRPM`.
2. The output of `pkggraph.rpmsProvidedBySRPM` is an array of package names. These can be fed almost directly into `rpmrepocloner.WhatProvides`.
3. Currently `WhatProvides` requires `*pkgjson.PackageVer` as input, but it uses it only to extract the package name and the optional version constraints. Here I'd suggest re-naming `WhatProvides` to `WhatProvidesPkgVer` and make `WhatProvides` accept a string (which would be either the package name or a package name along with version constraints). The re-named `WhatProvidesPkgVer` would be a simple wrapper around the new `WhatProvides`.

This should allow us to use existing logic and simplify the changes to this file.

Reply:
nodesProvidesBySRPM is more useful than rpmsProvidedbySRPM because node contains a lot more information than just the SRPM path.

Reply:
I looked closer at where `isSRPMAvailableUpstream` is used and I see that my initial suggestion was over-complicated. It looks like for the cycle resolution we only care if the currently analyzed run node from the cycle is available upstream. For that purpose, I believe we can completely skip `NodesProvidedBySRPM` and go straight to calling `cloner.WhatProvides(node.VersionedPkg)`. This way we can simplify the code significantly. We would only need the node and the cloner as input.

Let me know, what you think, and if I missed something important.

Reply:
You're saying that only the rpm of the single run node in the cycle is enough to break the cycle. Not all the rpms provided by that package are required. This should work.